Backfilling scheduler part 2

benarmston opened the pull request:
This PR is the second pass at a backfilling scheduler.  It is a significant rewrite of the firstaattempt with the following changes in behaviour.

1. Backfilling can only happen if a job has a run time limit, either defined by the user or using the default from the partition.

2. Backfilling works correctly for non-homogenous nodes on a partition.


## Algorithm

The algorithm used is based on the algorithm described in http://docs.adaptivecomputing.com/maui/8.2backfill.php. With the following changes:

1. Backfill windows are implicitly modelled.
2. The fit algorithm is a "quietest node" algorithm.

I'll describe the algorithm here in brief(ish).

When the backfilling scheduler schedules jobs for a partition, it iterates over the jobs and attepts to resources to each job.  If successful it moves onto allocate the next job.  This is so far idential to the behaviour of the FIFO scheduler.  If the scheduler cannot allocate resources to a job it attempts to create a reservation for that job.

To create a reservation, the scheduler looks at all nodes with sufficient resources to run the candidate job.  It then looks at the jobs running on the node and orders them by their guaranteed end time.  The scheduler then determines at what point in time the node will have sufficient resources available to run the candiate job assuming its current jobs stop in the order previously determined.  This gives us the time at which we can guarantee that the node will be available to run the candidate job.

Once we have such an end time for all nodes capable of running the candidate job we can calculate the earliest point at which we can guarantee there will be sufficient resources to run the candidate job.

Using those "availability times", we create a reservation.  Selecting the nodes that will allow the candidate job to run at the earliest time possible.

Where there is a choice of two nodes to add to the reservation, either of which would not change the guaranteed start time.  We prefer the node which is not currently able to run the job over one that can.  This is an attempt to allow as much backfilling as possible.

Once the reservation is created, we continue examining the remaining jobs and see if any can be allocated resources without using the resources that are part of the reservation.

If a backfill candidate cannot be allocated, we continue considering the remaining jobs, until we have either considered all jobs or we have reached the limit of the number of jobs to consider whilst allocating resources.


## Allocation vs Reservation

Reservations are similar to allocations, but they differ in a couple of respects.

 1. An allocation is fixed.  Once it has been added to the allocation registry we are committed to it.

 2. A reservation is ephemeral.  Once it is created, the only part of it we are committed to is that the start time will not be delayed.  The start time could be brought forward, and the selected nodes could change entirely.


# Limitations

If a node belongs to multiple partitions, the behaviour of reservations for that node is not defined.  I don't currently, know what the desired behaviour would be; does one partition respect the reservations made by another partition?  Do reservations (or partitions) have a priority order, such that higher priority partitions can overrule lower reservations from lower priority partitions?  A reservation is to be made on the highest priority job.  Is priority global or is it per partition?  How does this interact with partition priorities if such things exist?  These are questions for a later time.



# Future work

## Better fit algorithm

Currently, when selecting which resource can be allocated to a job, we chose the node that can run the job the most number of times, i.e., a "quietest fit" algorithm.  This fit algorithm is used both for the initial allocation of jobs and for backfilling jobs.

We use a slightly different algorithm for determining which nodes should be added to a reservation.  That algorithm is detailed above.

We should add alternative fit algorithms, such as a "first fit" and a "best fit" algorithm and allow them to be configurable.  These algorithms should be used for the initial allocation, backfilling and reservation creation.

Implementing this is likely to require the "allocation windows" mentioned in the linked paper.

## Create a configurable number of reservations

Currently, a single reservation for the highest priority job is created.  We should allow creating reservations for the highest N priority jobs.

benarmston committed:
First pass at non-homogeneous backfilling specs

Given the test data there is no backfilling.  Changing the nodes to have
more than 1 CPU (and a different number of CPUs) and changing the jobs
to require more than 1 CPU (and a different number) will open
opportunities backfilling.

benarmston committed:
Refactor scheduler specs

This makes adding new, more complicated, specs for the backfilling
scheduler easier.

benarmston committed:
Failing specs that need to be made to work

When these pass, we'll have a good backfilling scheduler.

benarmston committed:
Replace break from nested loop with throw/catch

The mechanics are relatively unchanged, but the code documents its
intention much more clearly.

benarmston committed:
Backfilling implementation

benarmston committed:
Backfilling specs are more tightly focused

benarmston committed:
Final specs to cover more cases

benarmston committed:
Refactor to make code clearer

WilliamMcCumstie commented on lib/flight_scheduler/load_balancer.rb:
```diff
@@ -27,14 +27,30 @@
     end
 
     def allocate
+      reservations = reservations()
```
AFAICS this is creating a local `reservations` variable that is shadowing the `reservations` method, which makes it hard to follow what is happening

I have a query about the `reservation()` function itself, so this issue will likely be resolved in due course.

benarmston replied:
Good point.  I think this was a hold over from an earlier implementation.

WilliamMcCumstie commented on lib/flight_scheduler/load_balancer.rb:
```diff
@@ -58,6 +74,22 @@ def allocate
 
     private
 
+    def reservations
```
In general I would expect the `reservations` method to be the same as `@reservations` as it is a noun, however this method is also applying a filter.

There are two possible solutions to this:
* Apply the filter within the `initialize` method, AFAICS the other reservations aren't used within this class and can be discarded
* Name the method to reflect that it preforms work, possible: `select_priority_reservations`

benarmston replied:
I've done something along those lines in https://github.com/openflighthpc/flight-scheduler-controller/pull/47/commits/b250ed883e5cebe89f82496bf8ec8af1bfa5bbe2

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -29,45 +29,59 @@
 
-  # Stores data for the backfilling algorithm.
+  # Represents a reservation of some nodes for a job.
+  class Reservation < Struct.new(:job, :start_time, :end_time, :nodes)
```
I'm not sure if a `Reservation` is a separate concept to an `Allocation`. They seem to be preforming similar tasks with the major distinction:
* `allocations` have started and are tracked by the `AllocationRegistry`, and
* `reservations` have not started and are tracked by the scheduler

Currently this is required because the `Job` tracks its `start_time` and `end_time` which means the `Reservation` needs to add them as additional fields.

If the `start_time`/`end_time` was moved into the `allocation`, I see no reason why `reservations`/`allocations` can't be collapsed into a single class.

benarmston replied:
`allocations` have started and are essentially immutable.  Its start time doesn't alter and neither do the nodes allocated to it.  Until they are unallocated.

`reservations` on the other hand are rather more ephemeral.  The start time is a "guaranteed latest start time".  As jobs are cancelled or finish early, the reservation could be recalculated with a different start time and a different set of nodes.

I made an attempt to move `Job#start_time` and `Job#end_time` into `Allocation` as that seems to be a better home for it.  However, `LoadBalancer#reservations` currently calculates the end time for the candidate job, i.e., a job that has not yet been allocated.  It does that in order to filter the reservations to those that are relevant for that job.  In time I expect that `AllocationRegistry#max_parallel_per_node` will learn the difference between an allocation and a reservation and the filtering done in  `LoadBalancer#reservations` will become unnecessary.  At that point I could see `Allocation` gaining  start time and end time.  Though collapsing reservations and allocations into the same class does not seem correct to me.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -29,45 +29,59 @@
+      if jobs.empty?
+        Time.now
+      else
+        jobs.map(&:end_time).max
```
This time calculation only holds for exclusive jobs. The node could have unused shared resources which aren't being used

benarmston replied:
Here `jobs` is a set of jobs that will (1) free up enough resources to enable the candidate to run on the node and (2) contains the jobs that have the earliest end time.

This set is calculated in the methods `jobs_in_completion_order` and `create_reservation` (when `availabilities` is calculated).

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -88,59 +102,109 @@ def run_allocation_loop(candidates)
+  def jobs_in_completion_order(node)
+    jobs_ordered_by_end = FlightScheduler.app.allocations.for_node(node.name)
+      .map { |alloc| alloc.job }
+      .reject { |j| j.time_limit.nil? }
```
Rejecting jobs without a timelimit looks a bit odd. I would have expected them to be sorted to the bottom of the list

benarmston replied:
Sorting to the bottom of the list would be wrong.  I'm using this method to determine the time at which sufficient jobs on the node have completed to allow the candidate to run.  In order to determine that only jobs with a guaranteed end time can be included in the calculation.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -88,59 +102,109 @@ def run_allocation_loop(candidates)
-    min_node_release = allocated_jobs.map do |job|
-      available_without_job = candidate.partition.nodes.select do |n|
-        alloc_reg.max_parallel_per_node(candidate, n, excluding_job: job) > 0
+    if availabilities.empty?
```
This availability check is being done before the additional unused nodes are added. Should it be done after?

benarmston replied:
Perhaps.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -88,59 +102,109 @@ def run_allocation_loop(candidates)
+
+      extra_nodes = potential_nodes
+        .select { |node| FlightScheduler.app.allocations.for_node(node.name).empty? }
+        .take(candidate.min_nodes - availabilities.length)
```
What happens if their is insufficient extra nodes to fulfil the reservation? Shouldn't this cause the reservation to fail?

benarmston replied:
Probably.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -29,45 +29,59 @@
-        next
-      end
-
       allocation = allocate_job(candidate)
```
I'm not sure if this is still valid. On the first iteration the `reservations` will be an empty array, however this may not be true subsequently.

This line probably needs to be `allocate_job(candidate, reservations: reservations)`

benarmston replied:
Good catch.  The line is still valid, but more by accident than design as we currently never make another iteration of the loop once a reservation has been created.  Fixed in https://github.com/openflighthpc/flight-scheduler-controller/pull/47/commits/dad9dbcf71a4500a24ed453fa283adeeaa8ee677

WilliamMcCumstie reviewed (commented):
These are my initially comments about the new back-filing scheduler. Possible more to come

benarmston committed:
Rename some methods/variables in load balancer

benarmston committed:
Fix bug in the making before its made

Currently, we never make another iteration of the allocation loop once a
reservation is created.  This will likely change in the future which
would introduce a bug with the previous code.

benarmston commented:
Following IRL discussion with @WilliamMcCumstie there are also the following changes to look at:

 * [x] Reservations should bias to as many unallocated nodes as possible.  This will allow the candidate to start sooner than biasing to as many allocated nodes as possible.
 * [x] If many nodes become available to run the candidate at the same time (e.g., there is an allocation on many nodes), we should prefer those nodes to unallocated nodes.  This allows the "backfill windows" to be larger without delaying the reservation.  We may want a fudge time here.

benarmston committed:
Correctly deserialisation persisted job start time

benarmston committed:
Fix backfilling and jobs without time limits

benarmston committed:
Refactor reservation finding

benarmston committed:
Spec to cover existing reservation bias

benarmston committed:
Implement desired bias

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -29,45 +29,211 @@
+    def jobs_waited_on
+      @jobs_waited_on ||=
+        begin
+          if FlightScheduler.app.allocations.for_node(node.name).empty?
```
This part of the conditional can be removed if `jobs_as_they_complete` first yields and empty array.

This has the added advantage that the `alloc_reg.max_parallel_per_node` check is preformed even if the job has no allocations. This has two advantages:
1. Allows allocated nodes with sufficient spare resources to run the `candidate` without waiting AND
2. Provides a firm guarantee that an IDLE node has sufficient resources

benarmston replied:
Good call.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -29,45 +29,211 @@
+    def availabilities_grouped_by_time_available
+      @availabilities_grouped_by_time_available ||=
+        begin
+          partitioned = {
```
Could we rename `partitioned` to `grouped` (or something to that affect)?

I know this method is logically "partitioning" the availabilities, but it has two problems:
1. It does not call the `partition` method on the `node_availabilities` array nor
2. Does it reflect the `Partition` models

For future readability/maintainability it would be clearer if it was named something different

benarmston replied:
Good call.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -29,45 +29,211 @@
+        Async.logger.debug("Unable to allocate candidate #{candidate.display_id}. Attempting to backfill.")
+        reservation = ReservationFinder.new(candidate).call
+        if reservation.nil?
+          Async.logger.debug("Unable to create reservation. Continuing normal allocation loop.")
```
AFAICS this conditional makes it possible to create a reservation for a non priority job?

Say there is a 10 node cluster: 2 to jobs with a timelimit and 6 without a timelimit. Then next job needs 5 nodes and will fail to create a `reservation` accordingly. The job after it needs 4 nodes and will reserve the 2 idle and 2 allocated with timelimit. The back-filling loop then runs as a NOOP and the scheduler exists.

This puts the scheduler in an unstable position because a time guarantee has been made on a non-priority job. If one of the 6 nodes without a time limit exists, it will trigger the scheduler to run again. As the reservations are not persisted, it will now successfully reserve the 3 idle and 2 timelimit nodes for the first job.

This breaks the guarantee made on the second job.

benarmston replied:
Good point.  As discussed IRL, if we encounter a job that cannot be reserved resources due to jobs running without a time limit, we do not backfill.  This ensures that the highest priority job is not delayed by backfilling.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/base_scheduler.rb:
```diff
@@ -150,20 +157,22 @@ def candidates(partition)
+              if next_task == previous_task
+                # We failed to schedule the ARRAY_TASK.  Move onto the next
+                # ARRAY_JOB or JOB.
+                throw :done_with_array_job
```
Is there a particular reason for using `catch`/`throw`?

AFAICS this could be done with `loop`/`break`

benarmston replied:
That's how I used to do it.  You complained that it was hard to follow due to the `loop`/`break` being nested inside of an each `each`.

WilliamMcCumstie replied:
Ah fair point. My initial comment was more highlighting the complexity of the process flow rather than the exact syntax.

Keep the `each`/`catch`/`loop`/`throw` for now as it works, eventually we may encapsulate the logic into another object.

WilliamMcCumstie reviewed (commented):
This is looking much better, just have a couple more comments

benarmston committed:
Prevent jobs without time limits breaking backfilling

If we encounter a candidate job that cannot be reserved because of jobs
running without a time limit.  We do not backfill.  This ensures that
the highest priority job is not delayed by backfilling.

benarmston committed:
Rename local variable: partitioned -> grouped

WilliamMcCumstie commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -244,4 +240,37 @@ def clear
+  def considering_allocations(node, given_allocations, excluding_jobs, reservations)
+    allocations = nil
+    if given_allocations
+      allocations = given_allocations
```
There should be a filter here to ensure the `node` is part of the `given_allocations`. This will be an easy mistake to make and something similar is already done for `reservations`.

benarmston replied:
Good catch.

WilliamMcCumstie commented on lib/flight_scheduler/allocation_registry.rb:
```diff
@@ -244,4 +240,37 @@ def clear
+    if given_allocations
+      allocations = given_allocations
+    else
+      allocations = @lock.with_read_lock { @node_allocations[node.name] }
```
Is the `LoadBalancer` used anywhere without an `allocations` input? If not this can probably be removed

benarmston replied:
It is.  So we still need this for now.

WilliamMcCumstie commented on lib/flight_scheduler/schedulers/backfilling_scheduler.rb:
```diff
@@ -29,45 +29,232 @@
+          if job_fits_partition?(candidate)
+            Async.logger.debug("Unable to create reservation. Exiting allocation loop.")
+            break
+          else
```
This conditional doesn't check directly if the job has a time-limit, instead it infers the condition from the fit algorithm. This _should_ work fine, however it has the possibility of hiding erroneous scheduling quirks.

Checking if the `candidate` is a fast operation so it might be worth splitting the logging condition. Something like:

```
if job_fit_partition?(candidate)
  ...
elsif candidate.timelimit?
  ...
else
  Async.logger.error "Unexpectedly failed to allocate the highest priority job"
end
```

benarmston replied:
The reservation can fail to be created if the candidate doesn't fit or if there are *other* jobs lacking a time limit.  A time limit on the candidate does not make any difference to whether a reservation can be created or not.  Unfortunately, checking if there are *other* jobs lacking a time limit is not necessarily quick.  You're right that this could hide a bug, but I don't think it will and we have sufficient logging to enable debugging (hopefully).

WilliamMcCumstie reviewed (approved):
Looking much better. Got a final few comments and should be good to go

benarmston committed:
Filter allocations
